hand-written.md

```diff
@@ -1,39 +1,32 @@
 
-### Source compatibility
+As usual for milestones, 2.12.0-M4 is not binary compatible with any other Scala release, including other 2.12 milestones. Scala 2.12 requires a Java 8 runtime.
 
-Scala 2.12 is mostly source compatible with 2.11.  Code that compiles on 2.11.x without deprecation warnings should compile on 2.12.x too, unless you use experimental APIs such as reflection.  If you find incompatibilities, please [file an issue](https://issues.scala-lang.org).
+### New features
 
-### Binary compatibility
+With M4, we consider 2.12.x feature complete. Please try it out while we still have time to fix any regressions in M5! For the next milestone, we'll focus on addressing any issues discovered with M4, overall robustness and polish, and some smaller improvements that are less likely to cause regressions or source incompatibilities. For RC1, we will try to remain binary compatible with M5, and we won't risk regressions except for the most critical bugs.
 
-Since Scala 2.11, minor releases of Scala are binary compatible with each other.
-Scala 2.12 will continue this tradition: every 2.12.x release will be binary compatible with 2.12.0.
-Milestone releases and release candidates, however, are **not** binary compatible with any other release.
+#### Trait compiles to an interface
+With Java 8 allowing concrete methods in interfaces, Scala 2.12 is able to compile a trait to a single interface. Before, a trait was represented as a class that held the method implementations and an interface. Note that the compiler still has quite a bit of magic to perform behind the scenes, so that care must be taken if a trait is meant to be implemented in Java. (Briefly, the following features require synthetic code in real classes that extend traits that do any of the following: defining fields, calling super, trait body statements, extending a class, relying on linearization to find implementations in the right super trait.)
 
-Scala 2.12 is not and will not be binary compatible with the 2.11.x series.  This allows us to keep improving the Scala compiler and standard library.  We are working with the community to ensure that core projects in the Scala eco-system become available for 2.12.  Please refer to this growing [list of libraries and frameworks](https://github.com/scala/make-release-notes/blob/2.12.x/projects-2.12.md).
+#### Java 8-style lambdas
 
-The [Scala 2.11.1 release notes](http://scala-lang.org/news/2.11.1) explain in more detail on how binary compatibility works in Scala.  The same policies apply to 2.12 as well.
+Scala 2.12 emits closures in the same style as Java 8, whether they target a FunctionN class from the standard library or a user-defined Single Abstract Method type. The type checker accepts a function literal as a valid expression for either kind of function type (built-in or SAM). This improves the experience of using libraries written for Java 8 in Scala.
```

Comment:
s/using libraries written for Java 8 in Scala/writing Scala code that uses Java 8 libraries/